### PR TITLE
fix websocket routine exiting

### PR DESCRIPTION
**Pull request by IPFSMain-Official**
Many times my miner encountered **websocket routine exiting** error, but later I found that it was because my lotus node was disconnected for a few minutes due to the network. Although the network was connected, the API always reported **websocket routine exiting**, and it was because the client triggered timeout(30s) to exit handleWsConn. Therefore, it would be better **not** to exit handleWsConn for timeout and wait for reconnection than to manually restart miner every time this happens.

### CODE REVIEW

**Comment by Xib1uvXi** [comment]
@magik6k This PR can effectively solve the RPC client error: sendRequest failed: websocket routine exiting error in miner

in our miner, ``Websocket routine exiting error`` in miner often appears:

```log
 2020-11-25T14:45:36.037+0800    ^[[34mINFO^[[0m events  events/events.go:104    restarting listenHeadChanges
2020-11-25T14:45:36.037+0800    ^[[34mINFO^[[0m events  events/events.go:104    restarting listenHeadChanges
2020-11-25T14:45:36.037+0800    ^[[31mERROR^[[0m        events  events/events.go:95     listen head changes errored: listenHeadChanges ChainNotify call failed: RPC client error: sendRequest failed: websocket routine exiting
2020-11-25T14:45:36.038+0800    ^[[31mERROR^[[0m        events  events/events.go:95     listen head changes errored: listenHeadChanges ChainNotify call failed: RPC client error: sendRequest failed: websocket routine exiting
2020-11-25T14:45:37.038+0800    ^[[34mINFO^[[0m events  events/events.go:104    restarting listenHeadChanges
2020-11-25T14:45:37.038+0800    ^[[34mINFO^[[0m events  events/events.go:104    restarting listenHeadChanges
2020-11-25T14:45:37.038+0800    ^[[31mERROR^[[0m        events  events/events.go:95     listen head changes errored: listenHeadChanges ChainNotify call failed: RPC client error: sendRequest failed: websocket routine exiting
2020-11-25T14:45:37.038+0800    ^[[31mERROR^[[0m        events  events/events.go:95     listen head changes errored: listenHeadChanges ChainNotify call failed: RPC client error: sendRequest failed: websocket routine exiting
2020-11-25T14:45:38.038+0800    ^[[34mINFO^[[0m events  events/events.go:104    restarting listenHeadChanges
2020-11-25T14:45:38.038+0800    ^[[31mERROR^[[0m        events  events/events.go:95     listen head changes errored: listenHeadChanges ChainNotify call failed: RPC client error: sendRequest failed: websocket routine exiting
2020-11-25T14:45:38.038+0800    ^[[34mINFO^[[0m events  events/events.go:104    restarting listenHeadChanges
2020-11-25T14:45:38.038+0800    ^[[31mERROR^[[0m        events  events/events.go:95     listen head changes errored: listenHeadChanges ChainNotify call failed: RPC client error: sendRequest failed: websocket routine exiting
2020-11-25T14:45:39.038+0800    ^[[34mINFO^[[0m events  events/events.go:104    restarting listenHeadChanges
2020-11-25T14:45:39.038+0800    ^[[34mINFO^[[0m events  events/events.go:104    restarting listenHeadChanges
2020-11-25T14:45:39.038+0800    ^[[31mERROR^[[0m        events  events/events.go:95     listen head changes errored: listenHeadChanges ChainNotify call failed: RPC client error: sendRequest failed: websocket routine exiting
2020-11-25T14:45:39.038+0800    ^[[31mERROR^[[0m        events  events/events.go:95     listen head changes errored: listenHeadChanges ChainNotify call failed: RPC client error: sendRequest failed: websocket routine exiting
2020-11-25T14:45:40.020+0800    ^[[31mERROR^[[0m        miner   miner/miner.go:176      failed to get best mining candidate: RPC client error: sendRequest failed: websocket routine exiting
2020-11-25T14:45:40.038+0800    ^[[34mINFO^[[0m events  events/events.go:104    restarting listenHeadChanges
2020-11-25T14:45:40.038+0800    ^[[34mINFO^[[0m events  events/events.go:104    restarting listenHeadChanges
2020-11-25T14:45:40.038+0800    ^[[31mERROR^[[0m        events  events/events.go:95     listen head changes errored: listenHeadChanges ChainNotify call failed: RPC client error: sendRequest failed: websocket routine exiting
2020-11-25T14:45:40.038+0800    ^[[31mERROR^[[0m        events  events/events.go:95     listen head changes errored: listenHeadChanges ChainNotify call failed: RPC client error: sendRequest failed: websocket routine exiting
2020-11-25T14:45:41.038+0800    ^[[34mINFO^[[0m events  events/events.go:104    restarting listenHeadChanges
```

The reason is that at some time miner and lotus will time out due to network fluctuations

```log
2020-11-25T14:45:35.019+0800    ^[[31mERROR^[[0m        rpc     go-jsonrpc@v0.1.2-0.20201008195726-68c6a2704e49/websocket.go:654        Connection timeout      {"remote": "127.0.0.1:1234"}
```


Related issues: 
https://github.com/filecoin-project/lotus/issues/4816
https://github.com/filecoin-project/lotus/issues/4762
https://github.com/filecoin-project/lotus/issues/3940

File: websocket.go [hunk 1]
```diff
@@ -652,7 +652,8 @@ func (c *wsConn) handleWsConn(ctx context.Context) {
 			}
 			c.writeLk.Unlock()
 			log.Errorw("Connection timeout", "remote", c.conn.RemoteAddr())
-			return
+			// Maybe the server was just shut down for a while, and then it started up again.
+			continue
```
Comment (magik6k) [inline websocket.go:656]: Are you sure that c.incoming will actually not hang / return an error (and make us reconnect) when we close the connection here? (even in cases when the server looses connection, then reboots, with no chance to cleanly close the TCP connection)

This is why this is was a return instead of a continue -> https://github.com/filecoin-project/go-jsonrpc/pull/23#discussion_r471565640

**Review by magik6k — COMMENTED** [review]
If we can prove that this doesn't have weird edge-cases where things would hang, this is good, but I'm not 100% sure

**Comment by whyrusleeping** [comment]
Would be great to have a test here...

File: websocket.go (hunk 1 above)
Comment (IPFSMain-Official) [reply websocket.go]: I think the problem is that both the client side and the server side use `handleWsConn`, but their strategies for handling connections should be slightly different

The client will only start handleWsConn once when NewClient called, if handleWsConn exit the client instance will become unavailable, rather than external NewClient again to get a new Client instance, but we should be looking forward to don't need to obtain a new instance Client again, it should be able to automatically reconnect, so client processing handleWsConn should have a principle: don't leave the function, but keep trying reconnection

The server starts a handleWsConn at ServeHTTP, and the client reconnects with a handleWsConn at every disconnect, so we need it to exit correctly, so the principle of handling handleWsConn on the server side is to close the connection and exit when the current connection is abnormal

Moreover, only the client has assigned a value of `connfactory`. Therefore, we can judge whether handlewsconn is called by the client or by the server, and implement the above two principles at a time

**Comment by IPFSMain-Official** [comment]
I tested it and changed it again, and it seems to work perfectly.
It will automatically reconnect as soon as the server starts after the client has timed out.
Server quitting `handleWsConn` correctly when the client disconnects.

File: websocket.go [hunk 2]
```diff
@@ -545,7 +545,10 @@ func (c *wsConn) handleWsConn(ctx context.Context) {
 		var timeoutCh <-chan time.Time
 		if timeoutTimer != nil {
 			if !timeoutTimer.Stop() {
-				<-timeoutTimer.C
+				select {
+				case <-timeoutTimer.C:
+				default:
```
Comment (whyrusleeping) [inline websocket.go:550]: adding a default here basically makes the select pointless. Should we just remove the timer altogether?

Comment (magik6k) [reply websocket.go:550]: This was doing the right thing (the select isn't needed). The channel needs to be drained when Stop returns false -> https://golang.org/pkg/time/#Timer.Stop

Comment (magik6k) [reply websocket.go:550]: Hrm, in local testing it.. apparently is needed
```
goroutine 130 [chan receive, 2 minutes]:
github.com/filecoin-project/go-jsonrpc.(*wsConn).handleWsConn(0xc00026e240, 0x346c880, 0xc000042848)
	/home/magik6k/gohack/github.com/filecoin-project/go-jsonrpc/websocket.go:548 +0x1208
created by github.com/filecoin-project/go-jsonrpc.websocketClient
	/home/magik6k/gohack/github.com/filecoin-project/go-jsonrpc/client.go:262 +0x2cf

```

Comment (magik6k) [reply websocket.go:550]: Ah, I see why it makes sense, `Stop()` will tell us that the timer has fired, but we may have already drained the channel below.
